compiler/optimizer/VPHandlers.cpp

```diff
@@ -7782,42 +7782,54 @@ TR::Node *constrainIand(OMR::ValuePropagation *vp, TR::Node *node)
                 (rhs->getLowInt() == TR::Compiler->cls.flagValueForArrayCheck(vp->comp())))
                {
-               TR::Node *firstGrandChild = firstChild->getFirstChild();
-               if ((firstGrandChild->getOpCodeValue() == TR::aloadi) &&
-                   (firstGrandChild->getSymbolReference() == vp->comp()->getSymRefTab()->findClassRomPtrSymbolRef()))
+               if (vp->trace())
+                  traceMsg(vp->comp(), "Found isArray test on node %p\n", node);
+
+               TR::Node *vftLoad = firstChild->getFirstChild();
+
+               if (vftLoad->getOpCodeValue() == TR::aloadi || vftLoad->getOpCodeValue() == TR::loadaddr)
                   {
-                  TR::Node *vftLoad = firstGrandChild->getFirstChild();
-                  if ((vftLoad->getOpCodeValue() == TR::aloadi) &&
-                      (vftLoad->getSymbolReference() == vp->comp()->getSymRefTab()->findVftSymbolRef()))
+                  TR::Node *baseExpression = NULL;
+                  if (vftLoad->getOpCodeValue() == TR::loadaddr)
+                     baseExpression = vftLoad;
+                  else
+                     baseExpression = vftLoad->getFirstChild();
+
+                  if (vp->trace())
+                     traceMsg(vp->comp(), "Base expression node for isArray test is %p\n", baseExpression);
+
+                  bool baseExpressionGlobal;
+                  TR::VPConstraint *baseExpressionConstraint = vp->getConstraint(baseExpression, baseExpressionGlobal);
+                  if (baseExpressionConstraint && baseExpressionConstraint->getClassType() && (baseExpressionConstraint->getClassType()->isArray() != TR_maybe))
                      {
-                     TR::Node *objectLoad = vftLoad->getFirstChild();
-                     bool objectLoadGlobal;
-                     TR::VPConstraint *objConstraint = vp->getConstraint(objectLoad, objectLoadGlobal);
-                     if (objConstraint && objConstraint->getClassType() && (objConstraint->getClassType()->isArray() != TR_maybe))
-                        {
-                        //dumpOptDetails(vp->comp(), "Folding isArray test in %p\n", objectLoad);
-                        //printf("Folding isArray test in %s\n", vp->comp()->signature()); fflush(stdout);
-                        if (objConstraint->getClassType()->isArray() == TR_yes)
-                           constraint = TR::VPIntConst::create(vp, rhs->asIntConst()->getLowInt()/*, isUnsigned*/);
-                        else
-                           constraint = TR::VPIntConst::create(vp, 0/*, isUnsigned*/);
-                        }
+                     if (vp->trace())
+                        traceMsg(vp->comp(), "Fold isArray test in %p\n", baseExpression);
+
+                     if (baseExpressionConstraint->getClassType()->isArray() == TR_yes)
+                        constraint = TR::VPIntConst::create(vp, rhs->asIntConst()->getLowInt()/*, isUnsigned*/);
```

Comment:
why is isUnsigned commented out?

Reply:
That is legacy code from the original implementation. TR::VPIntConst::create takes only 2 arguments now. Commit is modified to remote isUnsigned.